Enable entities (elems, nodes, etc) to be flagged during RB mesh pre-evaluation

apicard95 opened the pull request:
* This update introduces a generic map in RB that can store various sets of dof_id_types. This can be used to identify elems, nodes, components and apply a specific treatment to those entities during EIM assembly.
* The map is stored in VectorizedEvalInput and transferred to RBParametrizedFunctions during mesh pre-evaluation (all qp data).
* This map is then used to fill a second map that will store the same properties but only for EIM points. This second map is stored in RBEIMEvaluation and can be serialized and deserialized during the training.

Commit:
Enable entities (elems, nodes, etc) to be flagged during RB mesh pre-evaluation

* This update introduces a generic map in RB that can store various sets of dof_id_types.
This can be used to identify elems, nodes, components and apply a specific treatment
to those entities during EIM assembly.
* The map is stored in VectorizedEvalInput and transferred to RBParametrizedFunctions
during mesh pre-evaluation (all qp data).
* This map is then used to fill a second map that will store the same properties
but only for EIM points. This second map is stored in RBEIMEvaluation and can
be serialized and deserialized during the training.

Commit:
Switch rb_property_map object from unique_ptr to reference

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2535207/), step Generate coverage on ec0a853 wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2f82b</th>
      <th colspan=4>#3998 ec0a85</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>62.37%</td>
      <td>62.35%</td>
      <td>-0.02%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72655</td>
      <td>72654</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>43839</td>
      <td>43870</td>
      <td>+31</td>
      <td>32</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3998/coverage_diff)

[Full coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/3998/coverage)

## Warnings

- New new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

jwpeterson reviewed (approved):
This looks good to me, but I think I will give you a chance to do some internal testing before merging? Let me know if you prefer to just merge now.

Commit:
Add a setter for the rb_property_map stored in RBParametrizedFunction

Commit:
Switch to directly using RBParam.Function rb_property_map for storing mesh data

* Before we were using a temporary VectorizedEvalInput which required the
map to be moved after mesh pre-evaluation. Also VectorizedEvalInput was passed
as a const reference to vectorized_evaluate() which needed a const_cast to
modify it (design flaw).
* Now the rb_property_map containing mesh data is still stored in the
RBParametrizedFunction but directly filled in vectorized_evaluate() while the
rb_property_map from the VectorizedEvalInput is only read in vectorized_evaluate().
